### PR TITLE
Modifications to enable building with MinGW compilers

### PR DESCRIPTION
**Pull request by ptosco**
- added -DBOOST_SYSTEM_NO_DEPRECATED to definitions for MINGW builds
- modified a few #ifdef's to discriminate better between MINGW and MSVC
- added a ios::binary to Code/GraphMol/FileParsers/testMolSupplier.cpp
  (I already had to add some of those in the past to avoid the CR+LF issues on Windows)
- modified Code/GraphMol/FMCS/MaximumCommonSubgraph.cpp and
  Code/GraphMol/FMCS/FMCS.h because checking if ProgressCallback ==
  MCSProgressCallbackTimeout does not work with DLLs as the function pointer changes
  upon different calls, so it not reliable to store it as a parameter, while it is to store a nullptr
- builds under MinGW with -DRDK_USE_BOOST_SERIALIZATION=OFF and
  RDK_BUILD_THREADSAFE_SSS=OFF; I'll see if I can get it to work with
  RDK_BUILD_THREADSAFE_SSS=ON

### CODE REVIEW

**Comment by greglandrum** [comment]
This looks good to me. 
@bp-kelley : you mentioned on the mailing list that you might do something with mingw. Are you also ok with this patch?

**Comment by bp-kelley** [comment]
@ptosco Nicely done, a simpler fix than I expected.

Should we add mingw builds to appveyor?  I have a pull request for a more inclusive build matrix that I can revisit and try and add mingw to the windows builds.

File: Code/GraphMol/FMCS/FMCS.h [hunk 1]
```diff
@@ -116,7 +116,7 @@ struct MCSParameters {
         AtomTyper(MCSAtomCompareElements),
         BondTyper(MCSBondCompareOrder),
         CompareFunctionsUserData(0),
-        ProgressCallback(MCSProgressCallbackTimeout),
```
Comment (bp-kelley) [inline Code/GraphMol/FMCS/FMCS.h]: This is a good solution.  If you want to see the details of what is going on, take a deep breath and read this:

```
https://www.akkadia.org/drepper/dsohowto.pdf
```

If you like beer, have a lot of it ready :)

**Comment by ptosco** [comment]
@greglandrum @bp-kelley If you can wait a little bit before merging my PR, I will add some CMake magic to link against the MSVC runtime with the MinGW compiler, in order to enable multi-threaded builds.

File: Code/GraphMol/FMCS/FMCS.h (hunk 1 above)
Comment (ptosco) [inline Code/GraphMol/FMCS/FMCS.h]: Oh, you need something stronger than beer to swallow that :)

Comment (greglandrum) [inline Code/GraphMol/FMCS/FMCS.h]: just reading the abstract made my head hurt

**Comment by greglandrum** [comment]
I'll wait until you patch comes through.

**Comment by greglandrum** [comment]
Re: appveyor
We unfortunately can't do that. The compile time limits are already too tight; adding another build target would really break things.

Good CI on windows is likely to require a different solution than appveyor.

**Comment by ptosco** [comment]
PR completed.

File: Code/RDGeneral/CMakeLists.txt [hunk 2]
```diff
@@ -6,7 +6,29 @@ rdkit_library(RDGeneral
...
+    set(systemRootPath "$ENV{SystemRoot}")
+    string(REGEX REPLACE "\\\\" "/" systemRootPath "${systemRootPath}")
+    message(FATAL_ERROR "Please set MSVC_RUNTIME_DLL to the full path to a msvcrXXX.dll "
+      "Microsoft Visual C Runtime DLL appropriate for your system; the most likely path is "
+      "${systemRootPath}/System32, or ${systemRootPath}/SysWOW64 in case you are building "
+      "32-bit RDKit under a 64-bit Windows OS.")
+  endif()
+endif(MINGW)
+
+target_link_libraries(RDGeneral ${RDKit_THREAD_LIBS} ${MSVC_RUNTIME_DLL})
```
Comment (greglandrum) [inline Code/RDGeneral/CMakeLists.txt]: I can't currently test this, but it seems odd that this runtime is only linked into RDGeneral. Does that actually end up working for the other libraries too?

File: Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py [hunk 3]
```diff
@@ -375,16 +375,24 @@ def test8MultiThreadMultiConf(self):
             ee = ff.CalcEnergy()
             energies.append(ee)
 
-        mol = Chem.AddHs(Chem.MolFromSmiles("CC(C)(C)c(cc12)n[n]2C(=O)/C=C(N1)/COC"))
-        cids = rdDistGeom.EmbedMultipleConfs(mol,200,maxAttempts=30,randomSeed=100,numThreads=4)
+        mol2 = Chem.AddHs(Chem.MolFromSmiles("CC(C)(C)c(cc12)n[n]2C(=O)/C=C(N1)/COC"))
```
Comment (greglandrum) [inline Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py]: These changes (and the ones in testDGeomHelpers.cpp) are reasonable, but really aren't part of this PR.
Please keep PRs single topic in the future.

File: Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py [hunk 4]
```diff
@@ -375,16 +375,24 @@ def test8MultiThreadMultiConf(self):
...
 
-        self.assertTrue(lstEq(energies, nenergies,tol=1e-6))
+        self.assertTrue(lstEq(energies, nenergies, tol=5.0))
             
+        for cid in cids:
+            msd = 0.0
+            for i in range(mol.GetNumAtoms()):
+                msd += (mol.GetConformer().GetAtomPosition(i) \
+                    - mol2.GetConformer().GetAtomPosition(i)).LengthSq()
+            msd /= len(cids)
```
Comment (greglandrum) [inline Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py]: you probably mean to divide by the number of atoms here, not the number of conformers.

File: Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py (hunk 3 above)
Comment (ptosco) [inline Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py]: It wasn't laziness, I promise :)
I thought of keeping those bits together as the issue was raised by the MinGW gcc 32-bit compiler, but it's true it's more of a general issue. I just split the PR.

File: Code/RDGeneral/CMakeLists.txt (hunk 2 above)
Comment (ptosco) [inline Code/RDGeneral/CMakeLists.txt]: The reason the MSVC runtime needs to be linked to RDGeneral is that when multithreaded support is compiled in, LocalSwitcher.cpp needs _configthreadlocale(), which is only available as part of the MSVC runtime:

`#ifdef RDK_THREADSAFE_SSS
      _configthreadlocale(_ENABLE_PER_THREAD_LOCALE);`

So the linker complains about an undefined reference to _configthreadlocale() if the MSVC runtime is not linked to RDGeneral, but that does not affect the other libraries. All C++ tests are passed.

File: Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py (hunk 4 above)
Comment (ptosco) [inline Code/GraphMol/DistGeomHelpers/Wrap/testDistGeom.py]: Oops, of course, just fixed that. It will come in the 2nd PR.

**Comment by greglandrum** [comment]
Thanks Paolo

**Pull request merged**